text/0003-blockspace-regions.md

```diff
@@ -0,0 +1,252 @@
+
+### The Regions Pallet
+
+Regions will be managed within a "Regions" Pallet which exposes the following storage and `Call`s:
+
+```rust
+// Updated whenever regions are created, modified, or collected.
+storage map Regions: RegionId -> Region;
+
+// Updated when regions are created, transfered, or collected.
+// This is required for runtime APIs or other higher-level logic on collators to iterate the regions assigned
+// to a specific para.
+storage double_map RegionsPerPara: (ParaId, RegionId) -> ();
+
+// All functions are gated to permissioned origins, which are controlled by governance and intended to be assigned to system chains for managing regions.
+
+// Create a region. This is gated on allowed origins, e.g. a `RegionCreator` origin, set by governance.
+//
+// This inserts the region into the `RegionsPerPara` and `Regions` maps.
+//
+// If the region's `end` is `Some`, this sets up garbage collection logic for the region.
+fn create(Origin, Region) -> RegionId;
+
+// Update the end-point of a region. This is gated on allowed origins, e.g. a `RegionCreator` origin, set by governance.
+//
+// This is a no-op if the region's end is already `Some` or there is no region with that ID.
+// Otherwise, this sets up garbage-collection logic for the region. If `end` is in the past, `end` is set to the current block number.
+fn set_end(Origin, RegionId, end: u32);
+```
+
+This API is intended to be wrapped by logic implementing interfaces such as RFC-5.
+
+Created regions for a core MUST NOT exceed a combined block rate of 1 at any block.
+
+### Changes to backing/availability
+
+Regions are used to modify the behavior of the parachain backing/availability pipeline. The first major change is that parachain candidates submitted to the relay-chain in the `ParasInherent` will be annotated with the `RegionId` that they are intended to occupy. Validators and collators do the work of figuring out which blocks are assigned to which regions, lifting the burden of granular scheduling off of the relay chain.
+
+The **maximum consumption** of a region at any block number `now` is given by:
+`maximum_consumption(now, region) = (min(now, end) - start) * rate`
+If `end` is `None`, it is treated as infinite.
+
+The **minimum consumption** of a region at any block number `now` is given by:
+`minimum_consumption(now, region) = maximum_consumption(now - SCHEDULING_LENIENCE, region)`
+
+The **effective consumption** of a region at any block number `now` is the given by:
+`max(region.consumption, minimum_consumption(now, region))`
+
+A submitted candidate which uses `p: PartsOf57600` of the core's resources for a parachain P at a block B is accepted if:
+  * There is no candidate pending availability for B
+  * `effective_consumption(B, region) + p <= maximum` if `maximum` is `Some`
```

Comment:
This `maximum` is the total maximum of the region, in case it needs to be less than the implicit maximum given by end.

e.g. an on-demand purchase might have a high rate and a low maximum.